<a href="https://colab.research.google.com/github/acaropi/Curso-Ciencia-de-Datos-/blob/main/tarea_01_pandas_ingestion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Curso de Introducción a la Ciencia de Datos

**Ejercicio SF Salaries**
Utilizaremos el `SF Salaries Dataset` Kaggle. Simplemente siga y complete las tareas que se detallan a continuación en negrita. Las tareas serán cada vez más difíciles a medida que avance.

La idea principal es aplicar pandas para organizar informacion, manipularla, buscar información en el dataset


**Importe pandas como pd. e imprima la version que esta usando**

In [ ]:
# Copie el código aca

In [ ]:
import pandas as pd
print("versión de pandas:")
print(pd.__version__)



versión de pandas:
2.2.2


**leer el archivo Salaries.csv en un dataframe llamado salarios**

se encuentra en esta url y se puede leer directamente con pandas

https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv


In [ ]:
# Copie el código aca
import pandas as pd
url = "https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv"
salarios = pd.read_csv(url)
salarios.head()


,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Notes,Agency,Status
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,NaN,San Francisco,NaN
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,NaN,San Francisco,NaN
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,NaN,San Francisco,NaN
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,NaN,San Francisco,NaN
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,NaN,San Francisco,NaN


**Use el método `.info()` para encontrar cuantas Columnas de atributos hay y que tipo de datos tiene cada una.**

In [ ]:
# Copie el código aca
salarios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


- Segun la información del dataframe su tamaño es de *memory usage: 14.7+ MB*

- las columnas **Notes** y **Status** solamente tienen valores *NaN*, verifique que dice **0 non-null** osea 0 valores son diferentes a *NaN*
entonces todos los valores son *NaN* en esas columnas

**Elimine las columnas *Notes* y *Status* Recuerde usar el método de forma permanente (inplace), y luego visualice cuantas filas y columnas tiene luego de eliminar las columnas**

In [ ]:
# Copie el código aca
salarios.drop(columns=["Notes","Status"],inplace=True)

In [ ]:
salarios.shape

(148654, 11)

**¿Cual es la nueva información del dataframe?**

In [ ]:
# Copie el código aca

In [ ]:
salarios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  object 
dtypes: float64(6), int64(2), object(3)
memory usage: 12.5+ MB


- Se puede verificar que al borrar las columnas que no tenian solo *NaN* se disminuye el tamaño del dataframe a: *memory usage: 12.5+ MB*

**Observe los primeros 5 registros del DataFrame.**

In [ ]:
# Copie el código aca

In [ ]:
salarios.head(5)

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,San Francisco
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,San Francisco
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,San Francisco
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,San Francisco
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,San Francisco


**Observe los ultimos 5 registros del DataFrame.**

In [ ]:
# Copie el código aca

In [ ]:
salarios.tail(5)

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
148649,148650,Roy I Tillery,Custodian,0.0,0.0,0.00,0.0,0.00,0.00,2014,San Francisco
148650,148651,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148651,148652,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148652,148653,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148653,148654,Joe Lopez,"Counselor, Log Cabin Ranch",0.0,0.0,-618.13,0.0,-618.13,-618.13,2014,San Francisco


**¿Cuantas filas y columnas tiene el dataframe?**

In [ ]:
# Copie el código aca

In [ ]:
salarios.shape

(148654, 11)

**Observe los 10 registros que estan desde el indice 3000**

In [ ]:
# Copie el código aca

In [ ]:
salarios.iloc[3000:3010]

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
3000,3001,WILLIAM BOYLE,FIREFIGHTER,105934.65,18530.98,16389.60,NaN,140855.23,140855.23,2011,San Francisco
3001,3002,ALAN LIN,ELECTRONIC MAINTENANCE TECHNICIAN,98965.80,41504.40,380.00,NaN,140850.20,140850.20,2011,San Francisco
3002,3003,WILLIAM CONLEY,SERGEANT III (POLICE DEPARTMENT),130457.78,3131.42,7256.39,NaN,140845.59,140845.59,2011,San Francisco
3003,3004,ROBERT MCDONALD,"LIEUTENANT, FIRE DEPARTMENT",123104.99,1833.56,15897.68,NaN,140836.23,140836.23,2011,San Francisco
3004,3005,MICHAEL COSTELLO,POLICE OFFICER III,112401.08,7957.74,20466.93,NaN,140825.75,140825.75,2011,San Francisco
3005,3006,EDNA HO,MANAGER V,140821.37,0.00,0.00,NaN,140821.37,140821.37,2011,San Francisco
3006,3007,RODRIGO CONEL,REGISTERED NURSE,126120.00,3981.78,10714.21,NaN,140815.99,140815.99,2011,San Francisco
3007,3008,VICTOR HURTADO,FIREFIGHTER,105934.67,13112.76,21763.50,NaN,140810.93,140810.93,2011,San Francisco
3008,3009,DONALD LANDRY,TRANSIT POWER LINE WORKER,92861.50,39789.32,8135.86,NaN,140786.68,140786.68,2011,San Francisco
3009,3010,FRANK LEE,"INSPECTOR III, (POLICE DEPARTMENT)",130481.56,725.40,9579.58,NaN,140786.54,140786.54,2011,San Francisco



La columna *Benefits* tiene valores `NaN`, ¿Cuantos NaN tiene la columna *Benefits*? (TIP: Seleccione la Columna y vefique cuantos Na tiene `.isna()` luego sumelos `.sum()` )

In [ ]:
# Copie el código aca

In [ ]:
salarios["Benefits"].isna().sum()

np.int64(36163)

Al parecer los valores faltantes en la columna Benefits es por que no se le pagan beneficios entonces:

**Reemplace los valores *NaN* de la Columna *Benefits* por el valor de cero (Recuerde usar el método de forma inplace) y mostrar los primeros 10 elementos de la lista**

In [ ]:
# Copie el código aca
salarios["Benefits"].fillna(0, inplace=True)

/tmp/ipykernel_714/4075898614.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  salarios["Benefits"].fillna(0, inplace=True)


In [ ]:
salarios.head(10)

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,0.0,567595.43,567595.43,2011,San Francisco
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,0.0,538909.28,538909.28,2011,San Francisco
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,0.0,335279.91,335279.91,2011,San Francisco
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,0.0,332343.61,332343.61,2011,San Francisco
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,0.0,326373.19,326373.19,2011,San Francisco
5,6,DAVID SULLIVAN,ASSISTANT DEPUTY CHIEF II,118602.00,8601.00,189082.74,0.0,316285.74,316285.74,2011,San Francisco
6,7,ALSON LEE,"BATTALION CHIEF, (FIRE DEPARTMENT)",92492.01,89062.90,134426.14,0.0,315981.05,315981.05,2011,San Francisco
7,8,DAVID KUSHNER,DEPUTY DIRECTOR OF INVESTMENTS,256576.96,0.00,51322.50,0.0,307899.46,307899.46,2011,San Francisco
8,9,MICHAEL MORRIS,"BATTALION CHIEF, (FIRE DEPARTMENT)",176932.64,86362.68,40132.23,0.0,303427.55,303427.55,2011,San Francisco
9,10,JOANNE HAYES-WHITE,"CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)",285262.00,0.00,17115.73,0.0,302377.73,302377.73,2011,San Francisco


**¿Cual es la nueva información del dataframe?**

In [ ]:
# Copie el código aca

In [ ]:
salarios.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          148654 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  object 
dtypes: float64(6), int64(2), object(3)
memory usage: 12.5+ MB


**Cuantos valores unicos hay en las columnas categoricas?**

In [ ]:
# Copie el código aca

In [ ]:
salarios.select_dtypes(include="object").nunique()

,0
EmployeeName,110811
JobTitle,2159
Agency,1


- la columna **Agency** solamente tiene un valor unico, entonces no aporta informacion, por lo tanto se puede eliminar

**Elimine la columna *Agency* del dataframe**

In [ ]:
# Copie el código aca


In [ ]:
salarios.drop(columns=["Agency"], inplace=True)

**Cual es el promedio de BasePay ?**

In [ ]:
# Copie el código aca

66325.4488404877

**Cual es la cantidad mas alta de OvertimePay en el dataset ?**

In [ ]:
# Copie el código aca

In [ ]:
salarios["BasePay"].mean()

np.float64(66325.4488404877)

**Realice una descripcion estadistica de las columnas numericas?**

In [ ]:
# Copie el código aca
salarios.describe()


,Id,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year
count,148654.000000,148045.000000,148650.000000,148650.000000,148654.000000,148654.000000,148654.000000,148654.000000
mean,74327.500000,66325.448840,5066.059886,3648.767297,18924.232839,74768.321972,93692.554811,2012.522643
std,42912.857795,42764.635495,11454.380559,8056.601866,17165.279495,50517.005274,62793.533483,1.117538
min,1.000000,-166.010000,-0.010000,-7058.590000,-33.890000,-618.130000,-618.130000,2011.000000
25%,37164.250000,33588.200000,0.000000,0.000000,0.000000,36168.995000,44065.650000,2012.000000
50%,74327.500000,65007.450000,0.000000,811.270000,23214.020000,71426.610000,92404.090000,2013.000000
75%,111490.750000,94691.050000,4658.175000,4236.065000,33468.980000,105839.135000,132876.450000,2014.000000
max,148654.000000,319275.010000,245131.880000,400184.250000,96570.660000,567595.430000,567595.430000,2014.000000


**En este dataset las variables `Id` y `Year` representan variables categorica, convertir estas columnas a su formato correcto**

In [ ]:
# Copie el código aca

In [ ]:
salarios["Id"] = salarios["Id"].astype("category")
salarios["Year"] = salarios["Year"].astype("category")

**la variable `Year` es categorica ordinal, primero detecte cuales son los valores unicos que tiene la columna**

In [ ]:
# Copie el código aca

In [ ]:
salarios["Year"].unique()

[2011, 2012, 2013, 2014]
Categories (4, int64): [2011, 2012, 2013, 2014]

**Apartir de la información anterior convierta la columna `Year` en variable categorica ordinal y visualice la columna**

In [ ]:
# Copie el código aca

In [ ]:
orden_year = [2011, 2012, 2013, 2014]

salarios["Year"] = pd.Categorical(
    salarios["Year"],
    categories=orden_year,
    ordered=True
)
salarios["Year"].head(10)

,Year
0,2011
1,2011
2,2011
3,2011
4,2011
5,2011
6,2011
7,2011
8,2011
9,2011


**Realice nuevamente una descripcion estadistica de las columnas numericas?**

In [ ]:
# Copie el código aca

In [ ]:
salarios.describe()

,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits
count,148045.000000,148650.000000,148650.000000,148654.000000,148654.000000,148654.000000
mean,66325.448840,5066.059886,3648.767297,18924.232839,74768.321972,93692.554811
std,42764.635495,11454.380559,8056.601866,17165.279495,50517.005274,62793.533483
min,-166.010000,-0.010000,-7058.590000,-33.890000,-618.130000,-618.130000
25%,33588.200000,0.000000,0.000000,0.000000,36168.995000,44065.650000
50%,65007.450000,0.000000,811.270000,23214.020000,71426.610000,92404.090000
75%,94691.050000,4658.175000,4236.065000,33468.980000,105839.135000,132876.450000
max,319275.010000,245131.880000,400184.250000,96570.660000,567595.430000,567595.430000


**¿Cuál es el título de trabajo de JOSEPH DRISCOLL? Nota: Utilice mayúsculas, de lo contrario, puede obtener una respuesta que no coincida.**

In [ ]:
# Copie el código aca

In [ ]:
salarios[salarios["EmployeeName"] == "JOSEPH DRISCOLL"]["JobTitle"]

,JobTitle
24,"CAPTAIN, FIRE SUPPRESSION"


**¿Cuanto gana JOSEPH DRISCOLL (incluyendo beneficios)?**

In [ ]:
# Copie el código aca

In [ ]:
salarios.loc[
    salarios["EmployeeName"] == "JOSEPH DRISCOLL",
    "TotalPayBenefits"
].values[0]

np.float64(270324.91)

**¿Cual es el nombre de la persona con mejor pago (incluyendo beneficios)?**

In [ ]:
# Copie el código aca

In [ ]:
salarios.loc[
    salarios["TotalPayBenefits"].idxmax(),
    "EmployeeName"
]

'NATHANIEL FORD'

**¿Cual es el nombre de la persona peor paga (incluyendo beneficios)? ¿Notan algo extraño sobre cuánto le pagan?**

In [ ]:
# Copie el código aca

In [ ]:
salarios.loc[
    salarios["TotalPayBenefits"].idxmin(),
    ["EmployeeName", "TotalPayBenefits"]
]

,148653
EmployeeName,Joe Lopez
TotalPayBenefits,-618.13


**Cuantos registros tienen un valor de TotalPayBeneficts menor que cero**

In [ ]:
# Copie el código aca

4

**¿Cual es el promedio (mean) BasePay de todos los empleados por año? los años van entre (2011-2014)?** (Tip: Usar groupby)

In [ ]:
# Copie el código aca

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

**¿Cuantos titulos de trabajos (JobTitle) unicos existen?**

In [ ]:
# Copie el código aca

2159

**¿Cuales son los 5 trabajos mas comunes?**

In [ ]:
# Copie el código aca

Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

**¿Cuántos puestos de trabajo Unicos existieron en 2013? (Ejemplo, los títulos de trabajo con solo una ocurrencia en 2013?)**

In [ ]:
# Copie el código aca

202

**¿Cuántas personas tienen la palabra Chief en el título de su trabajo?** Apoyese en los métodos para manipular strings y en el método `.apply()`

In [ ]:
# Copie el código aca

627

Este código es valido, se recomienda hacer el código mas simple y agregar typing y documentación.

**Bono: ¿Existe una correlación entre la longitud del nombre del Título del trabajo y el Salario?**

crea una nueva columna donde se guarde la longitud del titulo del trabajo. tip: usar el método `.apply()`

In [ ]:
# Copie el código aca

0         46
1         31
2         31
3         36
4         44
          ..
148649     9
148650    12
148651    12
148652    12
148653    26
Name: title_len, Length: 148654, dtype: int64

Utilice el método `.corr()` de Pandas para calcular la correlacion

In [ ]:
# Copie el código aca

,title_len,TotalPayBenefits
title_len,1.000000,-0.036878
TotalPayBenefits,-0.036878,1.000000
